src/object/srv_obj.c

```diff
@@ -2978,3 +3023,837 @@ obj_verify_bio_csum(crt_rpc_t *rpc, daos_iod_t *iods,
+			D_GOTO(out, rc = 0);
+
+		/* Abort it firstly if exist but with different (old) epoch,
+		 * then re-execute with new epoch.
+		 */
+		if (rc == -DER_MISMATCH)
+			rc = vos_dtx_abort(ioc->ioc_vos_coh, DAOS_EPOCH_MAX,
+					   &dcsh->dcsh_xid, 1);
+
+		if (rc < 0 && rc != -DER_NONEXIST)
+			D_GOTO(out, rc);
+	}
+
+	rc = dtx_begin(ioc->ioc_coc, &dcsh->dcsh_xid, &dcsh->dcsh_epoch,
+		       dcde->dcde_write_cnt, oci->oci_map_ver,
+		       &dcsh->dcsh_leader_oid, NULL, 0, dcsh->dcsh_mbs, &dth);
+	if (rc != 0)
+		goto out;
+
+	rc = ds_obj_dtx_handle_one(rpc, dcsh, dcde, dcsr, ioc, &dth);
+	rc = dtx_end(&dth, ioc->ioc_coc, rc);
+
+out:
+	D_CDEBUG(rc != 0, DLOG_ERR, DB_IO,
+		 "Handled DTX "DF_DTI" on non-leader: "DF_RC"\n",
+		 DP_DTI(&dcsh->dcsh_xid), DP_RC(rc));
+
+	return rc;
+}
+
+static int
+obj_obj_dtx_leader(struct dtx_leader_handle *dlh, void *arg, int idx,
+		   dtx_sub_comp_cb_t comp_cb)
+{
+	struct ds_obj_exec_arg	*exec_arg = arg;
+	struct daos_cpd_args	*dca = exec_arg->args;
+	int			 rc = 0;
+
+	/* handle local operation */
+	if (idx == -1) {
+		if (!(exec_arg->flags & ORF_RESEND)) {
+			struct daos_cpd_disp_ent	*dcde;
+
+			dcde = ds_obj_cpd_get_dcde(dca->dca_rpc,
+						   dca->dca_idx, 0);
+			/* The check for read capa has been done before handling
+			 * the CPD RPC. Here, only need to check the write capa.
+			 */
+			if (dcde->dcde_write_cnt != 0) {
+				rc = obj_capa_check(dca->dca_ioc->ioc_coh,
+						    true);
+				if (rc != 0) {
+					comp_cb(dlh, idx, rc);
+
+					return rc;
+				}
+			}
+
+			rc = ds_obj_dtx_handle_one(dca->dca_rpc,
+				ds_obj_cpd_get_dcsh(dca->dca_rpc, dca->dca_idx),
+				dcde,
+				ds_obj_cpd_get_dcsr(dca->dca_rpc, dca->dca_idx),
+				dca->dca_ioc, &dlh->dlh_handle);
+		}
+
+		if (comp_cb != NULL)
+			comp_cb(dlh, idx, rc);
+
+		return rc;
+	}
+
+	/* Dispatch CPD RPC and handle sub requests remotely */
+	return ds_obj_cpd_dispatch(dlh, arg, idx, comp_cb);
+}
+
+static int
+ds_obj_dtx_leader_prep_handle(struct daos_cpd_sub_head *dcsh,
+			      struct daos_cpd_sub_req *dcsrs,
+			      struct daos_shard_tgt *tgts, uint32_t tgt_cnt,
+			      uint32_t req_cnt, uint32_t *flags)
+{
+	int	rc = 0;
+	int	i;
+
+	for (i = 0; i < req_cnt; i++) {
+		struct daos_cpd_sub_req		*dcsr;
+		struct daos_cpd_update		*dcu;
+
+		dcsr = &dcsrs[i];
+		if (dcsr->dcsr_opc != DCSO_UPDATE)
+			continue;
+
+		dcu = &dcsr->dcsr_update;
+		if (dcu->dcu_iod_array->oia_oiods == NULL)
+			continue;
+
+		rc = obj_ec_rw_req_split(dcsr->dcsr_oid, dcu->dcu_iod_array,
+					 dcsr->dcsr_nr, dcu->dcu_start_shard,
+					 dcu->dcu_ec_tgts, dcsr->dcsr_ec_tgt_nr,
+					 tgt_cnt, tgts, &dcu->dcu_ec_split_req);
+		if (rc != 0) {
+			D_ERROR("obj_ec_rw_req_split failed for obj "
+				DF_UOID", DTX "DF_DTI": "DF_RC"\n",
+				DP_UOID(dcsr->dcsr_oid),
+				DP_DTI(&dcsh->dcsh_xid), DP_RC(rc));
+			break;
+		}
+
+		if (dcu->dcu_ec_split_req != NULL)
+			*flags |= DRF_HAS_EC_SPLIT;
+	}
+
+	return rc;
+}
+
+static void
+ds_obj_dtx_leader_ult(void *arg)
```

Comment:
Just curious, will this be any group of object operation or standard daos object opc.   if it is standard object opc, why not reuse ds_obj_xx_handler().

what is the plan for current ds_obj_xxx_handler()? I was hopping we can unify the process of the cpd and single normal rpc.  but the patch seems separate it.   or I miss sth here.

Reply:
It is not the regular ds_obj_xx_handler(). Because even if single DTX in the CPD RPC, it also can contains multiple modification. We need special handling such case, such as update_begin() firstly for all sub-requests, then bulk for everyone if needed, and then update_end() one by one. The case will be more complex if the DTX contains both update and punch. In a shot time, it is not easy to unify the process for CPD RPC and standalone update/punch RPC. But we can consider it the future.

Reply:
ok, I see. let's check later.